sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -925,18 +928,22 @@ public boolean start() throws IOException {
       consumer = source.consumerFactoryFn.apply(source.consumerConfig);
       consumer.assign(source.assignedPartitions);
 
-      // seek to consumedOffset + 1 if it is set
       for (PartitionState p : partitionStates) {
-        if (p.consumedOffset >= 0) {
-          LOG.info("{}: resuming {} at {}", name, p.topicPartition, p.consumedOffset + 1);
-          consumer.seek(p.topicPartition, p.consumedOffset + 1);
+        if (p.consumedOffset >= 0 || p.consumedOffset == EMPTY_PARTITION_CONSUMED_OFFSET) {
+          consumer.seek(p.topicPartition, Math.max(p.consumedOffset + 1, 0L));
         } else {
-          LOG.info("{}: resuming {} at default offset", name, p.topicPartition);
+          // set consumed offset to (next offset - 1), otherwise checkpoint would contain invalid
+          // offset until we read first record from this partition.
+          long nextOffset = consumer.position(p.topicPartition);
+          p.consumedOffset = nextOffset == 0
+              ? EMPTY_PARTITION_CONSUMED_OFFSET : nextOffset - 1;
         }
+
+        LOG.info("{}: reading from {} at offset {}", name, p.topicPartition, p.consumedOffset + 1);
```

Comment:
This will log: `${readerName}: reading from ${topic} at offset -99` for a read attempt of an empty partition.
Maybe use `nextOffset` instead of `p.consumedOffset + 1` or log different messages for resuming and initialising.